# Analysis of the results

## Metrics analysis of word2vec networks

Italian

In [4]:
f = open('./word2vec_it/it_w_graphf.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

transitivity: 0.02446800001
assortativity: -0.10927938516889217
number_vertices: 795908
clusters_n: 2840
number_edges: 94102813
w_connected: False
average degree: 236.467
density: 0.00029710325320549293


Russian

In [3]:
f = open('./word2vec_rus/rus_graphf.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

transitivity: 0.45140185987949577
assortativity: 0.2895434379522005
number_vertices: 63415
clusters_n: 5014
number_edges: 333285
w_connected: False
average degree: 10.511
density: 0.000165755756017643


As we can see, Italian graph has more vertices and more edges than Russian, because Italian word2vec model had more tokens than Russian.  Surprisingly, these two networks which were built similarly and from similar kinds of sources have different signs in the measure of assortativity. We assume that this happened because of the different tresholds chosen for this two networks (cosine distance). For Italian network it was 0.65, for Russian -- 0.52. 
Transitivity of Russian network is much higher than in Italian network. We expected word2vec graphs to have high global clustering coefficients (transitivity), thus, more connected triangles, because if word A is a (quasi)synonym of B, B is a (quasi)synonym of C, it is logical that A is going to be synonymous to C. If the treshold in Italian were lower, most probably, the number of completed triangles in the network would be higher and, thus, transitivity. 
Another interesing difference between two networks is their average degree. The average degree in Italian network is 23 times higher than in Russian. Anomally big quantity of high-degree vertices in Italian network causes this effect.

![title](word2vec_it/it_w_degreedistr.png)

![title](word2vec_rus/ru_w_degreedistr.png)

The slight reflection of this tendency can also be noticed in Russian network.
We will find out which words in particular causes this peculiarity.

In [27]:
# Russian
data = g.degree()
in_values = sorted(set(data))
in_hist = {x: g.degree().count(x) for x in in_values}

In [28]:
for j in sorted(in_hist, key = in_hist.get, reverse = True):
    if j > 10 and j < 100 and in_hist[j] > 100 and  in_hist[j] < 1000:
        print(str(j) + '; ' + str(in_hist[j]))

11; 959
12; 848
13; 758
14; 654
15; 609
16; 605
17; 489
18; 446
20; 394
19; 391
21; 358
23; 326
22; 309
24; 305
25; 254
26; 229
28; 223
27; 219
29; 194
50; 194
30; 187
49; 175
51; 175
32; 174
31; 167
35; 151
33; 149
34; 148
40; 141
52; 140
39; 135
36; 134
38; 132
53; 131
37; 116
48; 116
54; 110
46; 108
59; 107
47; 104
43; 102
45; 102
56; 101


Nodes with degree = 49-51 stand out from the ascending order of degrees (first column).

In [30]:
print(', '.join(g.vs.select(_degree = 49)["name"]))
print('###')
print(', '.join(g.vs.select(_degree = 50)["name"]))
print('###')
print(', '.join(g.vs.select(_degree = 51)["name"]))

мэнэ_S, хиба_S, жулинский_S, ильген_S, горынь_S, вайоминг_S, калифорний_S, джорджия_S, нью-гемпшир_S, трутневой_A, сетунь_S, шуба_S, хлопонин_S, починок_S, брусенский_A, антониевый_A, акимовка_S, станиславовна_S, альмон_S, антониев_S, гидроксид_S, дудинцева_S, аронов_S, белашев_S, мыга_S, дзюбай_S, бавильский_A, худенький_A, ярко-желтый_A, коричневатый_A, корнилов_S, узкоплечий_A, бордовый_A, сопение_S, принстонский_A, агоев_S, чебенеев_S, хандыга_S, голынок_S, степанович_S, издатинлит_S, берзиня_S, несторович_S, слаповский_S, каргатский_A, минаевич_S, клементьич_S, рыжеватый_A, овраг_S, безносик_S, сидорченко_S, ульяновский_A, шарки_S, серейский_S, амударья_S, миркес_S, куваев_S, лещина_S, битов_S, облеков_S, рахманова_S, перикова_S, раиса_S, днепр_S, катерина_S, деменьшин_S, юхновская_S, демидович_S, курбановский_S, спарк_S, реп_S, сидякин_S, толстая_S, анисимовна_S, тепт_S, судзуха_S, немпт_S, немо_S, обтяжка_S, голлер_S, эгиз_S, хохлик_S, смиттен_S, пальчевский_S, флокс_S, деркул_S

The words with degree = 49-51 are for the most part Russian names. 

Italian
>>> print(', '.join(g.vs.select(_degree = 1037)["name"]))
avernas_NOM, kamberi_NOM, kvilekval_ADJ, nomeidae_ADJ, gambelii_NOM, kaibing_NOM, lesecce_NOM, repetition_NOM, fishes_ADJ, thrifty_ADJ, sentino_ADJ, rafsanjan_NOM, aeternam_NOM, addereto_NOM, coronatorum_NOM, verzera_NOM, olonne_NOM, jurisdictionibus_NOM
>>> print('###')
###
>>> print(', '.join(g.vs.select(_degree = 1230)["name"]))
tuatuari_ADJ, plugari_NOM, peasants_NOM, varemeist_ADJ, raconteurs_ADJ, padden_ADJ, temben_ADJ, meidassa_ADJ, aedif_NOM, narcissist_ADJ, marquel_NOM, frostuntil_NOM, copertas_NOM, eicas_ADJ, nozzae_ADJ, lemke_NOM
>>> print('###')
###
>>> print(', '.join(g.vs.select(_degree = 1102)["name"]))
edmontonensis_NOM, lacurne_ADJ, dienste_NOM, abstinence_NOM, monoko_NOM, chajbulaev_NOM, giurdignano_NOM, excellentibus_NOM, immunit_NOM, cimadera_NOM, margerie_NOM, kipner_NOM, creidhne_NOM, muina_ADJ, simplicibus_ADJ, naiùn_NOM, reichtum_NOM, chivré_NOM
>>> 

The abnormality in Italian graph is caused by latin words (like *immunit*), geographical designations (*Cimadera*), names (*Verzera*), and words from other languages (*fishes*).
To sum up, both Italian and Russian networks have a massive amount of peripheral words which became high-degree nodes in the network due to the network architecture algorithm. The high treshold which was establish because of computational capacity limitations made low frequency words get highest degrees, as the cosine similarity between them is higher than between the most similar common words. Compare the most similar words for *нефтегазкомплектсервис* (a name of a firm) and *абрикос* (apricot):<br>
нефтегазкомплектсервис крионорд 0.8124246597290039<br>
абрикос вишня 0.6917681694030762

## Top degree, top betweenness centrality in word2vec networks

Italian, nodes with the highest degrees

In [32]:
f = open('./word2vec_it/it_w_topdegree.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

56062;dekkhinathiri_NOM
56045;puiatu_ADJ
52933;pappírs_ADJ
49776;foreti_NOM
49456;bardusco_ADJ
46452;perfectionibus_NOM
45744;contestab_NOM
45306;mimophis_NOM
44608;gerip_NOM
43554;escrú_NOM
43114;kirna_NOM
42614;vaudefilms_NOM
42581;tabío_ADJ
41443;blastidi_NOM
40216;mxx_NOM
40212;ccclvi_NOM
40177;dortous_NOM
40002;holdi_NOM
39632;caponum_NOM
39529;syampati_NOM


Russian, nodes with the highest degrees

In [33]:
f = open('./word2vec_rus/ru_w_topdegree.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

371;кибалко_S
342;голяндин_S
321;клевчень_S
318;гажев_S
314;гоник_S
304;мосалев_S
301;харченков_S
299;щуплов_S
298;соколовер_S
293;цапелик_S
280;мельков_S
278;иванюк_S, убоженко_S
278;иванюк_S, убоженко_S
277;франюк_S
276;братерский_A, сливяк_S
276;братерский_A, сливяк_S
275;аркуша_S
274;рудиковский_S
273;баричев_S
272;костьев_S


Words which have the highest degrees in Russian and Italian word2vec networks are also low-frequency.

Italian, nodes with the highest betweenness centrality measure

In [48]:
f = open('./word2vec_it/it_w_topbetweennes.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

dekkhinathiri_NOM
foreti_NOM
bardusco_ADJ
foreti_NOM
pappírs_ADJ
escrú_NOM
meltada_ADJ
feuerstein_ADJ
cajum_NOM
stansfield_ADJ
ccclvi_NOM
mxx_NOM
syampati_NOM
caponum_NOM
holdi_NOM
blastidi_NOM
tabío_ADJ
mimophis_NOM
perfectionibus_NOM
teesta_NOM


Russian, nodes with the highest betweenness centrality measure

In [49]:
f = open('./word2vec_rus/ru_w_topbetweennes.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

голяндин_S
кибалко_S
харченков_S
гажев_S
клевчень_S
убоженко_S
соколовер_S
братерский_A,
щуплов_S
гоник_S
баричев_S
костьев_S
мосалев_S
мелийский_A
минеевич_S
комодо_S
тутечка_S
трутневой_A
гоник_S
сумленный_S


Words which have the highest betweenness centrality measures in Russian and Italian word2vec networks are also low-frequency.

## Metrics analysis of Google 2-Grams networks

Italian

In [11]:
f = open('./google_it/it_google_graphf.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

w_connected: False
assortativity: -0.08879987405502526
transitivity: 0.016367680663209917
number_vertices: 23067
clusters_n: 1462
number_edges: 49299
density: 0.00018531249456949138
average degree: 4.274


German

In [14]:
f = open('./google_de/de_google_graphf.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

w_connected: False
assortativity: -0.10766518979680163
transitivity: 0.008052953647246059
number_vertices: 27527
clusters_n: 1789
number_edges: 73563
density: 0.0001941723734963608
average degree: 5.345


Russian

In [13]:
f = open('./google_rus/google_graphf.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

density: 0.0008420803903874514
clusters_n: 164
transitivity: 0.05728370850350855
w_connected: False
assortativity: -0.09959788508259527
number_edges: 33100
average degree: 7.466
number_vertices: 8867


The output above shows that Italian an German networks are quite similar to each other, while Russian co-occurrence network slightly differs from them by some parameters. It is denser, has much less vertices and clusters, average degree and transitivity are slightly bigger. The total count of tokens in Russian, Italian, and German corpora is 67, 40, 64 milliard respectively. As the corpora volumes are compatible, the difference between network analysis metrics can be explained by the fact that lemmatisation was applied to Russian data while preprocessing, but to German and Italian networks was not. 
In general, the difference discussed above is not drastic, and networks exhibit similar properties.

67137666353 - Russian
40288809936 - Italian
64784628286 - German

## Scale-free?

As it can be seen in the figures showing degreee distribution of word2vec networks, degree distribution of both graphs does not follow power law. Thus, they cannot be considered scale-free.
Co-occurrence networks degree distribution, on the contrary, follows power law and can be called scale-free.

![title](google_de/de_g_degreedistr1.png)

![title](google_it/it_google_degreedistr.png)

![title](google_rus/ru_g_degreedistr.png)

## Small-world?

Only word2vec network of the Russian language shows global clustering coefficient to be characterized as having the small-world feature (0.45), the other networks have too small transtivity metric. 
Although Russian word2vec network is not connected, the average path can be counted for its giant component (largest subgraph), if it contains the majority of network nodes.

In [16]:
import igraph
g = igraph.Graph.Read_Ncol('./graphs/rus_word2vec_graph.ncol', directed=False, names=True).simplify(combine_edges=max)

In [17]:
community = g.community_multilevel(weights='weight')

In [21]:
print(community.giant().vcount())

2066


In [22]:
print(community.giant().vcount()/g.vcount()*100)

3.257904281321454


As it can be seen from the previous cell, the giant component amount only 3% from the number of vertices. Thus, the graph cannot be counted as a small-world.

## Top degree, top betweenness centrality in Google 2-Grams networks

Italian, nodes with the highest degrees

In [34]:
f = open('./google_it/it_google_topdegree.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

1081;sono
886;and
681;era
420;the, sia
420;the, sia
391;des
375;può
370;hanno
340;aveva
314;und
297;der
285;essere
244;deve
201;viene
194;que
183;generale
177;sguardo
175;two
174;own
166;sociale


Russian, nodes with the highest degrees

In [36]:
f = open('./google_rus/ru_google_topdegree.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

1253;быть
508;этот
453;мочь
409;человек
396;являться
339;тот
327;том
314;форма
300;жизнь
299;такой
285;часть
275;работа
273;язык
261;другой
258;день
251;тип
250;год
249;русский
242;характер
228;число



German, nodes with the highest degrees

In [37]:
f = open('./google_de/de_google_topdegree.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

4005;ist
1905;sind
1799;werden
1755;wird
1440;hat
1385;war
1039;wurde
811;haben
792;kann
576;hatte
569;sein
472;and
460;wurden
435;waren
421;können
296;sei
293;worden
285;gibt
263;Arbeit
243;Zeit


Top positions in degree ranking belong to functional words (*and, und, que*), modal verbs (*può, deve, мочь, kann, können*), verbs 'to be' (*ist, sind, was, sei, waren, быть, sono, era, essere, sia*), 'to have' (*hat, haben, hatte,  hanno, aveva*), 'to become' (*werden, word, wurde, worden*). They usually serves as auxilary verbs. The German word *gibt* (geben - to give) is used in the contruction 'there is(are)' -- *es gibt*. <br>
Above all this functional language units, in these lists there are also high-frequency words like *Zeit* (time), *человек* (a human/a man), *часть* (a part), *день* (a day/afternoon). Moreover, there are polysemious words:
<li>generale - general, main, in general, common, majority</li>
<li>sguardo - glare, glance, look, view </li>
<li>тип - type, kind, model, pattern, character </li>
<li>характер - character, nature, temper, personality </li>
<li>число - number, date, day, quantity </li>
<li>форма - form, shape, uniform, mould, outline, contour</li>
<li>Arbeit - work, job, labour. </li><br>
The output corresponds with Zipf's law and Zipf's law of meaning.

Italian, nodes with the highest betweenness centrality measure

In [43]:
f = open('./google_it/it_google_topbetweennes.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

sono
era
and
sia
des
aveva
und
que
the
può
hanno
Mr.
Oxford
ehe
der
est
von
G.B.
plus
P.M.


Russian, nodes with the highest betweenness centrality measure

In [45]:
f = open('./google_rus/rus_google_topbetweennes.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

быть
этот
том
человек
мочь
форма
тот
часть
тип
являться
язык
год
работа
день
группа
жизнь
е
характер
тьма
другой


German, nodes with the highest betweenness centrality measure

In [46]:
f = open('./google_de/de_google_topbetweennes.csv', 'r', encoding='utf-8')
for line in f:
    print(line.strip())

ist
sind
war
hat
and
werden
wird
wurde
Hrsg.
ff.
haben
kann
Uwe
Otto
hatte
Ulrich
Vladimir
GmbH
ccm
Hl.


In the top of betweenness centrality ranking there are also many words which serves as functional units in the language (articles, 'to be', 'to have', 'to become'), modal verbs. 
While preprocessing the Russian language corpora we had an opportunity not to include all the tokens from other languages by applying a filter to latin characters and  checking whether a token has been annotated by pymorphy2 as 'UNKN' or 'LATN'. As we did not apply such kind of preprocessing to German and Italian languages, several tokens from exterior languages appeared to be in the graph, and high-frequency of them were assigned with the high measure of betweenness centrality.

## Community detection analysis

The applied community detection algorithm has separated graphs into cluster in a very detailed manner. In the tables below you can see quantity of clusters for each languages and the number of clusters, in which there are >= 30 words. 

Word2Vec



|         |Italian|Russian|
|---------|-------|------ |
|n_cluster| 2840  | 125   |
| >= 30   | 5014  | 285   |



Google 2-Grams

|         |Italian|Russian|German|
|---------|-------|------ |------|
|n_cluster| 1462  | 164   | 1789 |
| >= 30   |   46  |  53   |  69  |

Clusters of 2 or 3 words, naturally, do not represent semantic domains. They usually contain collocations, names of entities, such as firms, geographical locations, abbreviations.
<li>шут гороховый</li>
<li>йоркширский терьер</li>
<li>Asthma bronchiale</li>
<li>Etica Nicomachea</li>
<li>freundlichen Grüßen (a German letter closing, 'best wishes')</li>

We are going to take a closer look at Google 2-grams graphs clusters which contain high-degree polysynonymous words (see the output above).

In [26]:
import igraph
g = igraph.Graph.Read_Ncol('./graphs/de_google_graphf.ncol', directed=False, names=True).simplify(combine_edges=max)

In [27]:
communityde = g.community_multilevel(weights='weight')

In [35]:
word = g.vs.find(name="Arbeit").index

In [78]:
for vertex in g.vs():
    vertex["label"] = vertex["name"]

In [36]:
for index, i in enumerate(communityde):
    for element in i:
        if element == word:
            print(index)
            cluster = communityde.subgraph(index)
            #igraph.plot(cluster, './google_it/sguardo.png', layout='drl')
            break

732


In [37]:
result = open('Arbeitcluster.txt', 'a', encoding='utf-8')
for element in cluster.vs():
    result.write(element["name"] + '\n')

In [7]:
n = open('sguardocluster.txt', 'r', encoding='utf-8').readlines()[:10]
print(''.join(n))

Studi
AA.W.
luogo
citato
ragione
seguito
Abbonamento
annuo
abbonamento
postale



![title](./google_it/sguardo.png)

This cluster is big and dense. Nevertheless, not all the vertices are related to 'sguardo' (glance, view) and we cannot claim that this cluster represents a semantic domain (*parola (word), dialogo (dialogue), dolore (pain), natura, problema* are in this cluster). 


The next cluster represents days of the week (*lunedi, sabato, domenica*), parts of the day (*sera (evening), mattino (morning)*, and other time markers (*ieri (yesterday), indomani (tomorrow, the next day)*). The central part of the graph represents 'time' domain. Nevertheless, there are some nodes which are not related to this domain. On th right end of the plot there are adjectives associated with *evening* (*sera*), for instance, *impensabile (unthinkable, impossible)*. On the left bottom of the figure some words non-Italian words and characters are presented.

![title](./google_it/pomodoro.png)

German clusters

We are going to investigate in which cluster word *Arbeit* is. 

In [98]:
for index, i in enumerate(communityde):
    for element in i:
        if element == word:
            print(index)
            cluster = communityde.subgraph(index)
            igraph.plot(cluster, './google_de/sguardo.png', layout='drl')
            break

732


In [13]:
n = open('Arbeitcluster.txt', 'r', encoding='utf-8').readlines()[:10]
print(''.join(n))

Aalto
and
Alvar
The
Aage
Brandt
have
Aarbog
for
Aarboger



![title](./google_de/sguardo.png)

This cluster also contains a lot of vertices and is dense. This cluster represents several semantic domains, as in contains abstract words, and some of them have different meanings. 

We will try to find a corresponding cluster of 'time' in the German co-occurrence network.

In [100]:
for vertex in g.vs():
    vertex["label"] = vertex["name"]
word = g.vs.find(name="Montag").index
for index, i in enumerate(communityde):
    for element in i:
        if element == word:
            print(index)
            cluster = communityde.subgraph(index)
            igraph.plot(cluster, './google_it/woche.png', layout='drl')
            break

658


![title](./google_de/woche.png)

As the cluster is too big and names of the vertices cannot be seen, we print out cluster words.

In [34]:
n = open('Zeitcluster.txt', 'r', encoding='utf-8').readlines()[:10]
print(''.join(n))


acht
gelassen
lassen
Acht
läßt
achten
ist
müssen
bleiben



There are many words which can co-occur with the days of the week in one sentence, but they cannot be related to the same domain: *gelernt (learned), kann (can), bleibt (stays)*

Russian co-occurrence network

There are no days of the week presented in Russian co-occurrence network (the frequency of biagrams was low, so that they were not included).

Russian word2vec network

In [106]:
for vertex in g.vs():
    vertex["label"] = vertex["name"]
for index, i in enumerate(communityrus):
    for element in i:
        if element == word:
            print(index)
            cluster = communityrus.subgraph(index)
            igraph.plot(cluster, './google_rus/lang1.png', layout='drl')
            break

104


## Conclusion

Google 2-Grams networks built in this research appeared to represent language with its semantic nature better than word2vec networks. However, we cannot claim that building a network based a word2vec model is not suitable for linguistic research, the algorithm for this type of network must be modified. The treshold should not be too high, future vertices (words) must be carefully selected by applying frequency dictionary for a model lexicon and/or collocation metrics (Dice coefficient, MI-score, LL-score and etc.) which will catch low frequency noise like bigrams with first and last names. Consequently, there will be no abnormalities in degree distribution and at the top of betweenness centrality and degree rankings will be represented some of the polysemious words. 
We expect word2vec networks to have much higher global clustering coefficient (like the Russian word2vec network we have built) than Google 2-Grams networks, because relationships of synonymy/antonymy are **transitive**, while co-occurrence relationships may not be transitive.
Our co-occurrence networks are scale-free, some words which have more than one meaning are high-degree, which corresponds with Zipf's law of meaning. Moreover, these words are primarily abstract, therefore, they connect vertices from different semantic domains and tend to be in huge clusters where more than one semantic domains are presented.

## Some plots of networks and clusters

Russian word2vec network

![title](word2vec_rus/drl_word2vec_rus.png)

Russian Google 2-Grams network

![title](google_rus/drl_google_rus.png)

German Google 2-Grams network with colored clusters

![title](google_de/drl_clusters_google_de.png)

Italian Google 2-Grams network with colored clusters

![title](google_it/drl_clusters_google_it.png)

Italian Google 2-Grams network

![title](google_it/drl_google_it.png)

Unfortunately, the Italian network was too large to be plotted.